nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/ArchiveDao.java

```diff
@@ -0,0 +1,114 @@
+    return jdbc.query(
+                    "select w.id id from nflow_workflow w, " +
+                    "(" +
+                    "  select parent.id from nflow_workflow parent " +
+                    "  where parent.next_activation is null and parent.modified <= ? " +
+                    "  and parent.root_workflow_id is null " +
+                    "  and not exists(" +
+                    "    select 1 from nflow_workflow child where child.root_workflow_id = parent.id " +
+                    "      and (child.modified > ? or child.next_activation is not null)" +
+                    "  )" +
+                    "  order by modified asc " +
+                    "  limit " + maxRows +
+                    ") as archivable_parent " +
+                    "where archivable_parent.id = w.id or archivable_parent.id = w.root_workflow_id",
+            new Object[] { toTimestamp(before), toTimestamp(before) }, new ArchivableWorkflowsRowMapper());
+  }
+
+  @Transactional
+  public void archiveWorkflows(List<Integer> workflowIds) {
+    String workflowIdParams = params(workflowIds);
+
+    archiveWorkflowTable(workflowIdParams);
+    archiveActionTable(workflowIdParams);
+    archiveStateTable(workflowIdParams);
+    deleteWorkflows(workflowIdParams);
+  }
+
+  private void archiveWorkflowTable(String workflowIdParams) {
+    String columns = columnsFromMetadata("nflow_workflow");
+    jdbc.update("insert into nflow_archive_workflow(" + columns + ") " +
+            "select " + columns + " from nflow_workflow where id in " + workflowIdParams);
+  }
+
+  private void archiveActionTable(String workflowIdParams) {
+    String columns = columnsFromMetadata("nflow_workflow_action");
+    jdbc.update("insert into nflow_archive_workflow_action(" + columns + ") " +
+            "select " + columns + " from nflow_workflow_action where workflow_id in " + workflowIdParams);
+  }
+
+  private void archiveStateTable(String workflowIdParams) {
+    String columns = columnsFromMetadata("nflow_workflow_state");
+    jdbc.update("insert into nflow_archive_workflow_state (" + columns + ") " +
+            "select " + columns + " from nflow_workflow_state where workflow_id in " + workflowIdParams);
+  }
+
+  private void deleteWorkflows(String workflowIdParams) {
+    jdbc.update("delete from nflow_workflow_state where workflow_id in " + workflowIdParams);
+    jdbc.update("update nflow_workflow set root_workflow_id=null, parent_workflow_id=null, parent_action_id=null " +
+            "where id in " + workflowIdParams);
+    jdbc.update("delete from nflow_workflow_action where workflow_id in " + workflowIdParams);
+    jdbc.update("delete from nflow_workflow where id in " + workflowIdParams);
+  }
+
+  private String columnsFromMetadata(String tableName) {
+    List<String> columnNames = jdbc.query("select * from " + tableName + " where 1 = 0", columnNamesExtractor);
+    return join(columnNames.toArray(), ",");
```

Comment:
no need for toArray conversion here? StringUtils join supports collections directly.